### PR TITLE
Convert system plugins to service provider

### PR DESCRIPTION
**Pull request by laoneo**
### Summary of Changes
Converts the system plugins to service providers.

### Testing Instructions
Do different things like login/logout on the front end, enable debug mode, play around with http headers.

### Actual result BEFORE applying this Pull Request
All works.

### Expected result AFTER applying this Pull Request
All works.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by Fedik** [comment]
@laoneo please keep plugin constructors consistent.
```php
new Plugin($dispatcher, $config, ... other variables)
```
instead of
```php
new Plugin(... other variables, $dispatcher, $config)
```
That would be nice :)

File: plugins/system/stats/layouts/field/data.php [hunk 1]
```diff
@@ -14,7 +14,7 @@
 use Joomla\CMS\Language\Text;
 
 /** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
-$wa = Factory::getApplication()->getDocument()->getWebAssetManager();
+$wa = $this->getApplication()->getDocument()->getWebAssetManager();
```
Comment (joomdonation) [inline plugins/system/stats/layouts/field/data.php]: Please revert this change. This is a form field layout, not run in plugin context, so you cannot use $this->getApplication()

File: plugins/system/languagefilter/languagefilter.xml [hunk 2]
```diff
@@ -9,8 +9,10 @@
 	<authorUrl>www.joomla.org</authorUrl>
 	<version>3.0.0</version>
 	<description>PLG_SYSTEM_LANGUAGEFILTER_XML_DESCRIPTION</description>
+	<namespace path="src">Joomla\Plugin\System\Languagefilter</namespace>
```
Comment (joomdonation) [inline plugins/system/languagefilter/languagefilter.xml]: `Languagefilter` need to be changed to `LanguageFilter`. Also, the file Languagefilter.php needs to be renamed to `LanguageFilter.php` . On Windows, I'm getting fatal error because class not found.

File: plugins/system/languagecode/languagecode.xml [hunk 3]
```diff
@@ -9,8 +9,10 @@
 	<authorUrl>www.joomla.org</authorUrl>
 	<version>3.0.0</version>
 	<description>PLG_SYSTEM_LANGUAGECODE_XML_DESCRIPTION</description>
+	<namespace path="src">Joomla\Plugin\System\Languagecode</namespace>
```
Comment (joomdonation) [inline plugins/system/languagecode/languagecode.xml]: For consistence, should we change `Languagecode` to `LanguageCode` ? Also, the class name and the filename, too.

File: plugins/system/languagefilter/src/Extension/Languagefilter.php [hunk 4]
```diff
@@ -90,34 +95,44 @@ class PlgSystemLanguageFilter extends CMSPlugin
...
+    private $languageFactory;
 
     /**
      * Constructor.
      *
-     * @param   object  &$subject  The object to observe
-     * @param   array   $config    An optional associative array of configuration settings.
+     * @param   DispatcherInterface   $dispatcher   The dispatcher
+     * @param   array                 $config       An optional associative array of configuration settings
+     * @param   UserFactoryInterface  $userFactory  The user factory
```
Comment (joomdonation) [inline plugins/system/languagefilter/src/Extension/Languagefilter.php]: Wrong docblock and missing definition for $app and $languageFactory here.

File: plugins/system/logout/src/Extension/Logout.php [hunk 5]
```diff
@@ -41,30 +36,33 @@ class PlgSystemLogout extends CMSPlugin
...
-     * @param   object  &$subject  The object to observe -- event dispatcher.
-     * @param   object  $config    An optional associative array of configuration settings.
+     * @param   DispatcherInterface      $dispatcher  The object to observe -- event dispatcher.
+     * @param   array                    $config      An optional associative array of configuration settings.
+     * @param   CMSApplicationInterface  $app         The object to observe -- event dispatcher.
      *
      * @since   1.6
      */
-    public function __construct(&$subject, $config)
+    public function __construct(DispatcherInterface $dispatcher, $config, CMSApplicationInterface $app)
```
Comment (joomdonation) [inline plugins/system/logout/src/Extension/Logout.php]: Please add `array` typehint for `$config` parameter

File: plugins/system/logrotation/logrotation.xml [hunk 6]
```diff
@@ -9,8 +9,10 @@
 	<authorUrl>www.joomla.org</authorUrl>
 	<version>3.9.0</version>
 	<description>PLG_SYSTEM_LOGROTATION_XML_DESCRIPTION</description>
+	<namespace path="src">Joomla\Plugin\System\Logrotation</namespace>
```
Comment (joomdonation) [inline plugins/system/logrotation/logrotation.xml]: Please change Logrotation here (and related code) to LogRotation.

File: plugins/system/schedulerunner/schedulerunner.xml [hunk 7]
```diff
@@ -9,12 +9,14 @@
 	<authorUrl>www.joomla.org</authorUrl>
 	<version>4.1</version>
 	<description>PLG_SYSTEM_SCHEDULERUNNER_XML_DESCRIPTION</description>
+	<namespace path="src">Joomla\Plugin\System\Schedulerunner</namespace>
```
Comment (joomdonation) [inline plugins/system/schedulerunner/schedulerunner.xml]: Please change **Schedulerunner** here (and related code) to **ScheduleRunner**

File: plugins/system/skipto/src/Extension/Skipto.php [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+                        // Selectors for landmark and headings sections
+                        'headings'  => 'h1, h2, h3',
+                        'landmarks' => 'main, nav, search, aside, header, footer, form',
+                    ],
+                ],
+            ]
+        );
+
+        /** @var Joomla\CMS\WebAsset\WebAssetManager $wa */
```
Comment (joomdonation) [inline plugins/system/skipto/src/Extension/Skipto.php]: Please change this to `/** @var \Joomla\CMS\WebAsset\WebAssetManager $wa */` . It is missing the backslash

File: plugins/system/stats/src/Extension/Stats.php [hunk 9]
```diff
@@ -6,19 +6,21 @@
...
- * @phpcs:disable PSR1.Classes.ClassDeclaration.MissingNamespace
  */
 
+namespace Joomla\Plugin\System\Stats\Extension;
+
+use Exception;
 use Joomla\CMS\Cache\Cache;
-use Joomla\CMS\Factory;
 use Joomla\CMS\Http\HttpFactory;
 use Joomla\CMS\Language\Text;
```
Comment (joomdonation) [inline plugins/system/stats/src/Extension/Stats.php]: Remove `use Joomla\CMS\Language\Text;` from there. It is not used anymore

File: plugins/system/webauthn/src/Authentication.php [hunk 10]
```diff
@@ -321,7 +321,7 @@ public function validateAttestationResponse(string $data): PublicKeyCredentialSo
 
         // Retrieve the stored user ID and make sure it's the same one in the request.
         $storedUserId = $this->session->get('plg_system_webauthn.registration_user_id', 0);
-        $myUser       = $this->app->getIdentity() ?? new User();
+        $myUser       = $this->getApplication()->getIdentity() ?? new User();
```
Comment (joomdonation) [inline plugins/system/webauthn/src/Authentication.php]: Please revert the change you made in this class. This is not the plugin class, the application is stored in $this->app property . `$this->getApplication()` is not available to use here.

**Review by joomdonation — COMMENTED** [review]
Thanks. There are few things left and I think this PR is ready.

**Comment by laoneo** [comment]
@joomdonation added your suggestions

**Comment by joomdonation** [comment]
I have tested this item :white_check_mark: successfully on 04cdf1ace149e6e82872685bee30ac075ba302bc<br /><br />Careful code review + tested as much as I can (both for mono language and multilingual website).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40216">issues.joomla.org/tracker/joomla-cms/40216</a>.</sub>

File: plugins/system/accessibility/services/provider.php [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+defined('_JEXEC') or die;
+
+use Joomla\CMS\Extension\PluginInterface;
+use Joomla\CMS\Factory;
+use Joomla\CMS\Plugin\PluginHelper;
+use Joomla\DI\Container;
+use Joomla\DI\ServiceProviderInterface;
+use Joomla\Event\DispatcherInterface;
+use Joomla\Plugin\Fields\Calendar\Extension\Calendar;
```
Comment (sandewt) [inline plugins/system/accessibility/services/provider.php]: use Joomla\Plugin\System\Accessibility\Extension\Accessibility;

File: plugins/system/accessibility/services/provider.php [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function register(Container $container): void
+    {
+        $container->set(
+            PluginInterface::class,
+            function (Container $container) {
+                $dispatcher = $container->get(DispatcherInterface::class);
+                $plugin     = new Calendar(
```
Comment (sandewt) [inline plugins/system/accessibility/services/provider.php]: new Accessibility;

File: plugins/system/accessibility/services/provider.php (hunk 11 above)
Comment (laoneo) [reply plugins/system/accessibility/services/provider.php]: Thanks!

File: plugins/system/privacyconsent/src/Extension/PrivacyConsent.php [hunk 13]
```diff
@@ -673,7 +659,7 @@ private function invalidateExpiredConsents()
             }
 
             $messageModel->notifySuperUsers(
-                Text::_('PLG_SYSTEM_PRIVACYCONSENT_NOTIFICATION_USER_PRIVACY_EXPIRED_SUBJECT'),
+                $this->getApplication()->getLanguage()->_('PLG_SYSTEM_PRIVACYCONSENT_NOTIFICATION_USER_PRIVACY_EXPIRED_SUBJECT'),
                 Text::sprintf('PLG_SYSTEM_PRIVACYCONSENT_NOTIFICATION_USER_PRIVACY_EXPIRED_MESSAGE', Factory::getUser($user->user_id)->username)
```
Comment (sandewt) [inline plugins/system/privacyconsent/src/Extension/PrivacyConsent.php:663]: Is there no `$this->getApplication()->getLanguage()->_` method yet for `Text::sprintf` ?

Comment (laoneo) [reply plugins/system/privacyconsent/src/Extension/PrivacyConsent.php:663]: Nope, because you can simply do `sprintf($this->getApplication()->getLanguage()->_(''), $args)` when the language string is in sprintf format.

Comment (sandewt) [reply plugins/system/privacyconsent/src/Extension/PrivacyConsent.php:663]: Thanks

Comment (sandewt) [reply plugins/system/privacyconsent/src/Extension/PrivacyConsent.php:663]: I'm still wondering why you didn't change line 663 to: 

`sprintf(this->getApplication()->getLanguage()_('PLG_SYSTEM_PRIVACYCONSENT_NOTIFICATION_USER_PRIVACY_EXPIRED_MESSAGE'), Factory::getUser($user->user_id)->username)`

Or did I miss something? 

[EDIT I hope I don't seem unfriendly. I'm just trying to better understand the code.]

Comment (laoneo) [reply plugins/system/privacyconsent/src/Extension/PrivacyConsent.php:663]: I just wanted to change as less logic as possible. Stuff like that can be done in a follow up pr.

Comment (sandewt) [reply plugins/system/privacyconsent/src/Extension/PrivacyConsent.php:663]: I can totally understand that.
